db/db_logical_block_size_cache_test.cc

```diff
@@ -0,0 +1,366 @@
+        cf_path_0_(dbname_ + "/cf_path_0"),
+        cf_path_1_(dbname_ + "/cf_path_1"),
+        kDirFds({
+          {dbname_, 0},
+          {data_path_0_, 10},
+          {data_path_1_, 11},
+          {cf_path_0_, 20},
+          {cf_path_1_, 21}
+        }) {
+    auto get_fd_block_size = [&](int fd) {
+      ncall_++;
+      return fd;
+    };
+    auto get_dir_block_size = [&](const std::string& dir, size_t* size) {
+      ncall_++;
+      *size = kDirFds.at(dir);
+      return Status::OK();
+    };
+    cache_.reset(new LogicalBlockSizeCache(
+        get_fd_block_size, get_dir_block_size));
+    env_.reset(new EnvWithCustomLogicalBlockSizeCache(
+        Env::Default(), cache_.get()));
+  }
+
+ protected:
+  std::string dbname_;
+  std::string data_path_0_;
+  std::string data_path_1_;
+  std::string cf_path_0_;
+  std::string cf_path_1_;
+
+  const std::map<std::string, int> kDirFds;
+
+  int ncall_ = 0;
+  std::unique_ptr<LogicalBlockSizeCache> cache_;
+  std::unique_ptr<Env> env_;
+};
+
+TEST_F(DBLogicalBlockSizeCacheTest, OpenClose) {
+  // Tests that Open will cache the logical block size for data paths,
+  // and Close will remove the cached sizes.
+  Options options;
+  options.create_if_missing = true;
+  options.env = env_.get();
+  options.db_paths = {{data_path_0_, 2048}, {data_path_1_, 2048}};
+
+  for (int i = 0; i < 2; i++) {
+    ncall_ = 0;
+    DB* db;
+    if (!i) {
+      printf("Open\n");
+      ASSERT_OK(DB::Open(options, dbname_, &db));
+    } else {
+      printf("OpenForReadOnly\n");
+      ASSERT_OK(DB::OpenForReadOnly(options, dbname_, &db));
+    }
+    // Logical block size of data paths are cached during Open.
+    ASSERT_EQ(2, ncall_);
+    ASSERT_EQ(10, cache_->GetLogicalBlockSize(data_path_0_ + "/sst", 100));
+    ASSERT_EQ(2, ncall_);
+    ASSERT_EQ(11, cache_->GetLogicalBlockSize(data_path_1_ + "/sst", 200));
+    ASSERT_EQ(2, ncall_);
+    ASSERT_OK(db->Close());
+    ASSERT_EQ(2, ncall_);
+    // Logical block size of data paths are removed from cache during Close.
+    ASSERT_EQ(100, cache_->GetLogicalBlockSize(data_path_0_ + "/sst", 100));
+    ASSERT_EQ(3, ncall_);
+    ASSERT_EQ(200, cache_->GetLogicalBlockSize(data_path_1_ + "/sst", 200));
+    ASSERT_EQ(4, ncall_);
+    delete db;
+  }
+}
+
+TEST_F(DBLogicalBlockSizeCacheTest, OpenDelete) {
+  // Tests that Open will cache the logical block size for data paths,
+  // and delete the db pointer will remove the cached sizes.
+  Options options;
+  options.create_if_missing = true;
+  options.env = env_.get();
+
+  DB* db;
+  ASSERT_OK(DB::Open(options, dbname_, &db));
+  // Logical block size of dbname_ is cached during Open.
+  ASSERT_EQ(1, ncall_);
+  ASSERT_EQ(0, cache_->GetLogicalBlockSize(dbname_ + "/sst", 100));
+  ASSERT_EQ(1, ncall_);
+
+  delete db;
+  ASSERT_EQ(1, ncall_);
+  // Logical block size of dbname_ is removed from cache during Close.
+  ASSERT_EQ(100, cache_->GetLogicalBlockSize(dbname_ + "/sst", 100));
+  ASSERT_EQ(2, ncall_);
```

Comment:
Can we validate in a more straight-forward way: adding a function in cache that determine whether it is empty, or return number of paths in the cache?

Reply:
Added an `IsEmpty()` method to the cache.

Reply:
And asserts that cache IsEmpty after every `delete db`

Reply:
Added `Size()` to return number of cached directories